File: backend/app/models/room.py
from datetime import datetime, UTC
from sqlalchemy import String, Integer, Boolean, DateTime, Enum as SQLEnum, Text, ForeignKey
from sqlalchemy.orm import Mapped, mapped_column, relationship
import enum

from app.core.database import Base

class RoomStatus(str, enum.Enum):
    ACTIVE = "active"
    INACTIVE = "inactive"

class Room(Base):
    __tablename__ = "rooms"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    capacity: Mapped[int] = mapped_column(Integer, nullable=False)
    location: Mapped[str] = mapped_column(String(255), nullable=False)
    building: Mapped[str | None] = mapped_column(String(100), nullable=True)
    floor: Mapped[str | None] = mapped_column(String(20), nullable=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    status: Mapped[RoomStatus] = mapped_column(SQLEnum(RoomStatus), default=RoomStatus.ACTIVE, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(UTC), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(UTC), onupdate=lambda: datetime.now(UTC), nullable=False)

    # Relationships
    images = relationship("RoomImage", back_populates="room", cascade="all, delete-orphan")
    equipment = relationship("RoomEquipment", back_populates="room", cascade="all, delete-orphan")
    bookings = relationship("Booking", back_populates="room")

class RoomImage(Base):
    __tablename__ = "room_images"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    room_id: Mapped[int] = mapped_column(Integer, ForeignKey("rooms.id"), nullable=False, index=True)
    image_path: Mapped[str] = mapped_column(String(500), nullable=False)
    is_primary: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(UTC), nullable=False)

    room = relationship("Room", back_populates="images")

class RoomEquipment(Base):
    __tablename__ = "room_equipment"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    room_id: Mapped[int] = mapped_column(Integer, ForeignKey("rooms.id"), nullable=False, index=True)
    name: Mapped[str] = mapped_column(String(100), nullable=False)

    room = relationship("Room", back_populates="equipment")

File: frontend/src/services/room.ts
import api from './api';

export interface RoomEquipment {
  id: number;
  name: string;
}

export interface RoomImage {
  id: number;
  image_path: string;
  is_primary: boolean;
}

export interface Room {
  id: number;
  name: string;
  capacity: number;
  location: string;
  building: string | null;
  floor: string | null;
  description: string | null;
  status: 'active' | 'inactive';
  created_at: string;
  updated_at: string;
  equipment: RoomEquipment[];
  images: RoomImage[];
}

export interface RoomCreateData {
  name: string;
  capacity: number;
  location: string;
  building?: string;
  floor?: string;
  description?: string;
  equipment?: string[];
}

export default {
  async getRooms(params: any = {}) {
    const response = await api.get('/rooms', { params });
    return response.data;
  },

  async getRoom(id: number) {
    const response = await api.get(`/rooms/${id}`);
    return response.data;
  },

  async createRoom(data: RoomCreateData) {
    const response = await api.post('/rooms', data);
    return response.data;
  },

  async updateRoom(id: number, data: any) {
    const response = await api.patch(`/rooms/${id}`, data);
    return response.data;
  },

  async deactivateRoom(id: number) {
    const response = await api.post(`/rooms/${id}/deactivate`);
    return response.data;
  },

  async activateRoom(id: number) {
    const response = await api.post(`/rooms/${id}/activate`);
    return response.data;
  },

  // Image Management
  async uploadImage(roomId: number, file: File) {
    const formData = new FormData();
    formData.append('file', file);
    const response = await api.post(`/rooms/${roomId}/images`, formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    });
    return response.data;
  },

  async deleteImage(roomId: number, imageId: number) {
    await api.delete(`/rooms/${roomId}/images/${imageId}`);
  },

  async setPrimaryImage(roomId: number, imageId: number) {
    const response = await api.post(`/rooms/${roomId}/images/${imageId}/set-primary`);
    return response.data;
  },
};
